### PR TITLE
Directly compress mip maps from Basis Universal

### PR DESCRIPTION
**Pull request by fire**
Part of: https://github.com/godotengine/godot-proposals/issues/4271

Discussed with @reduz:

Only support basisu with uastc.

It has been veto'd to not include all possible platforms at the same time. Do not store all of etc2, dxt5, bptc at the same time due to excess file size.

---

To increase efficiency enable compressed mip maps from Basis Universal.
To improve quality enable UASTC RDO compression.
To keep compatibility use the first mip of the previous internal Godot format.

This does a trick where I use only the first image from now on to read its mipmaps.

----

Part of the V-Sekai group. https://github.com/v-sekai

### CODE REVIEW

File: editor/import/resource_importer_texture.cpp [hunk 1]
```diff
@@ -380,7 +376,8 @@ void ResourceImporterTexture::_save_ctex(const Ref<Image> &p_image, const String
 
 	Ref<Image> image = p_image->duplicate();
 
-	if (((p_compress_mode == COMPRESS_BASIS_UNIVERSAL) || (p_compress_mode == COMPRESS_VRAM_COMPRESSED && p_force_po2_for_compressed)) && p_mipmaps) {
+	if ((((p_compress_mode == COMPRESS_BASIS_UNIVERSAL) && p_force_po2_for_compressed) || (p_compress_mode == COMPRESS_VRAM_COMPRESSED && p_force_po2_for_compressed)) &&
+			p_mipmaps) {
```
Comment (akien-mga) [inline editor/import/resource_importer_texture.cpp:380]:
```suggestion
	if (p_force_po2_for_compressed && p_mipmaps && ((p_compress_mode == COMPRESS_BASIS_UNIVERSAL) || (p_compress_mode == COMPRESS_VRAM_COMPRESSED))) {
```
Equivalent?

File: modules/basis_universal/register_types.cpp [hunk 2]
```diff
@@ -52,44 +54,47 @@ enum BasisDecompressFormat {
...
+		}
+		params.m_source_mipmap_images.push_back(source_images);
+		params.m_quality_level = basisu::BASISU_QUALITY_MIN;
+		params.m_mip_fast = true;
+		params.m_pack_uastc_flags = basisu::cPackUASTCLevelFaster;
 		params.m_multithreading = true;
-		//params.m_quality_level = 0;
-		//params.m_disable_hierarchical_endpoint_codebooks = true;
-		//params.m_no_selector_rdo = true;
+		params.m_quality_level = 0;
```
Comment (akien-mga) [inline modules/basis_universal/register_types.cpp:88]: You're setting it twice to different values.

**Review by akien-mga — COMMENTED** [review]
Only gave it a very cursory check, would need in-depth review by @reduz. But the changes seem sensible, if BU can now support mipmaps properly we should use it.

File: modules/basis_universal/register_types.cpp (hunk 2 above)
Comment (fire) [reply modules/basis_universal/register_types.cpp:88]: Resolving.

**Comment by fire** [comment]
@lyuma since you worked on the mipmaps code, can you review that area?

File: modules/basis_universal/register_types.cpp (hunk 2 above)
Comment (fire) [reply modules/basis_universal/register_types.cpp]: Thanks, you found the cause of the lower than expected performance! Thanks.

**Comment by fire** [comment]
Can be archived. I'll will merge it with https://github.com/godotengine/godot/pull/62499

**Comment by fire** [comment]
tl;dr

This pr does three things.

1. use one mipmap call to basisu instead of several
2. Discussed maybe separating basisu high quality and basisu fast
3. basisu high quality uses uastc and lowest quality
4. basisu fast uses etc1s and highest possible quality

**Comment by fire** [comment]
Superseded by: https://github.com/godotengine/godot/pull/62499
